### DISCUSSION TITLE
GetPatches method throws ERROR_MORE_DATA

### DISCUSSION THREAD

**Discussion by oshut** [wixtoolset/issues#8770] [Questions] [answered]
Hello everyone,
I'm getting ERROR_MORE_DATA exception when calling `GetPatches` from `PatchInstallation` class.
Error occurs in `NativeMethods.MsiEnumPatchesEx` because user SID doesn't fit in buffer.
Line 118:
![image](https://github.com/user-attachments/assets/13457ddc-d16f-45c6-b7be-1fcc5fc716ea)

Am I missing something or is there any way to workaround this? because I can't change how `NativeMethods.MsiEnumPatchesEx` behaves and I can't change SID either

Regards,
Ostap

**Comment by bevanweiss** [comment]
What does your code actually say?
Perhaps you should have put the code inside a code block instead of pasting a screen shot.

ERROR_MORE_DATA is telling you that the buffer size you provided was insufficient... 
https://learn.microsoft.com/en-us/windows/win32/api/msi/nf-msi-msienumpatchesexw

**Reply by oshut** [reply]
Hello @bevanweiss 
Sorry about screenshot
The code I've sent is not mine, it's inside WiX library and I didn't provide that buffer, I'm just calling `PatchInstalltion.GetPatches` inside of my code
Like this:
`PatchInstallation.GetPatches(null, productCode, "s-1-1-0", UserContexts.All, PatchStates.All);`

**Reply by oshut** [reply]
@bevanweiss is there anything else I can provide that might help?

**Reply by bevanweiss** [reply, marked as answer]
You could try rebuilding and testing WiX with a modification to the line below to increase this 40 characters to 256
![image](https://github.com/user-attachments/assets/b71e0469-0370-4abb-8d5a-9274efc1b72e)

If that does resolve it, then you could file an issue with the problem, and the suggestion to increase this.

**Reply by wmanning** [reply]
That doesn't limit the string buffer to 48 characters.  It will allocate more memory if needed.

https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.-ctor?view=net-8.0#system-text-stringbuilder-ctor(system-int32)

**Reply by bevanweiss** [reply]
> That doesn't limit the string buffer to 48 characters. It will allocate more memory if needed.
> 
> https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.-ctor?view=net-8.0#system-text-stringbuilder-ctor(system-int32)

I'm very confident this is NOT true across Native boundaries.  Prior to actually invoking the Native Win32api method the memory of the .NET is pinned in place, and cannot reallocate.  So if the Native method wants more, then it will fail (and return ERROR_MORE_DATA)

**Reply by barnson** [reply]
Also, every release of WiX has sources and pdbs, so anybody can run it under a debugger.
